code/core/src/phone/java/com/adobe/marketing/mobile/services/ui/notification/PushTemplateUtils.kt

```diff
@@ -0,0 +1,150 @@
+/*
+  Copyright 2024 Adobe. All rights reserved.
+  This file is licensed to you under the Apache License, Version 2.0 (the "License");
+  you may not use this file except in compliance with the License. You may obtain a copy
+  of the License at http://www.apache.org/licenses/LICENSE-2.0
+  Unless required by applicable law or agreed to in writing, software distributed under
+  the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
+  OF ANY KIND, either express or implied. See the License for the specific language
+  governing permissions and limitations under the License.
+*/
+
+package com.adobe.marketing.mobile.services.ui.notification
+
+import android.content.ContentResolver
+import android.content.Context
+import android.content.pm.PackageManager
+import android.net.Uri
+import androidx.annotation.VisibleForTesting
+import com.adobe.marketing.mobile.services.Log
+import com.adobe.marketing.mobile.services.ui.notification.templates.AEPPushTemplate
+import org.json.JSONArray
+import org.json.JSONException
+import org.json.JSONObject
+
+internal object PushTemplateUtils {
```

Comment:
This class has methods that don't necessarily fit in as extension functions but are used across templates. Happy to discuss a more suitable place to put these

Reply:
I feel like we
- `getIconWithResourceName` , `getDefaultAppIcon`, `getSoundUriForResourceName` can be grouped as  `AppResourceExtensions` or similar. 
- `isValidIcon`, `getActionButtonsFromString` and `getActionButtonFromJSONObject` as a file private methods in `NotificationCompatBuilderExtension`
let me know what you think,

Reply:
Like the idea of `AppResourceExtensions`! But for `getActionButtonsFromString` and `getActionButtonFromJSONObject`, it would be very hard to unit test them as private methods. How about adding it as a method to `AEPPushTemplate.ActionButton` class? `AEPPushTemplate` already have the `actionButtonString` anyway

Reply:
Also thinking it might make more sense to have `ActionButton` as a separate class and maintain a list of action buttons in AEPPushTemplate

Reply:
Given that`AEPPushTemplate ` is the base template for all other templates, it should only contain attributes that are common to all templates. I believe not all templates allow action buttons. So I prefer not to have the action button related logic (except for ActionButton model definition) there.

If testing is the only issue, why not expose them as `internal` and `@VisibleForTesting`? These will anyway be moved to `notification.internal` package when migrating to its own repo.